draft-ietf-mls-protocol.md

```diff
@@ -1698,85 +1684,110 @@ struct {
     HPKEPublicKey public_key;
     Credential credential;
     Capabilities capabilities;
+    Extension extensions<V>;
 
-    LeafNodeSource leaf_node_source;
-    select (leaf_node_source) {
+    ClientInfoSource client_info_source;
+    select (client_info_source) {
+        case key_package:
         case add:
             Lifetime lifetime;
-
         case update:
             struct {}
-
         case commit:
             opaque parent_hash<V>;
     }
 
-    Extension extensions<V>;
     // SignWithLabel(., "LeafNodeTBS", LeafNodeTBS)
     opaque signature<V>;
-} LeafNode;
+} ClientInfo;
 
 struct {
     HPKEPublicKey public_key;
     Credential credential;
     Capabilities capabilities;
-
-    LeafNodeSource leaf_node_source;
-    select (leaf_node_source) {
-        case key_package:
-            Lifetime lifetime;
-
-        case update:
-            struct{};
-
-        case commit:
-            opaque parent_hash<V>;
-    }
-
     Extension extensions<V>;
 
-    select (leaf_node_source) {
+    ClientInfoSource client_info_source;
+    select (client_info_source) {
         case key_package:
-            struct{};
-
+        case add:
+            Lifetime lifetime;
         case update:
             opaque group_id<V>;
-
         case commit:
+            opaque parent_hash<V>;
             opaque group_id<V>;
     }
-} LeafNodeTBS;
+} ClientInfoTBS;
+
+LeafNode ClientInfo;
 ~~~~~
 
-The `public_key` field conains an HPKE public key whose private key is held only
-by the member occupying this leaf.  The `credential` contains authentication
-information for this member, as described in {{credentials}}.
+The `public_key` field conains an ephemeral HPKE public key whose private key
+is held only by the issuer of the KeyPackage (when used inside a KeyPackage),
+or the member occupying the referenced leaf (when used as a LeafNode).
+The `credential` contains permanent or semi-permanent authentication
+information for this member or potential member, as described in {{credentials}}.
 
 The `capabilities` field indicates what protocol versions, ciphersuites,
 protocol extensions, and non-default proposal types are supported by a client.
 Proposal types defined in this document are considered "default" and thus need
-not be listed.  Extensions that appear in the `extensions` field of a LeafNode
+not be listed.  Extensions that appear in the `extensions` field of a ClientInfo
 MUST be included in the `extensions` field of the `capabilities` field.
 
-The `leaf_node_source` field indicates how this LeafNode came to be added to the
-tree.
+The `client_info_source` field indicates the context in which the ClientInfo
+object is used.
```

Comment:
This is incorrect.  The ClientInfo is used as a leaf in the tree for TreeKEM, regardless of how it got there.  This field says how it got there, thus `source`.

Reply:
The idea of how it got there being relevant even days later is a little weird, so adding an extra sentence or two explaining this would be useful. I started originally trying to do a lighter edit, and trying to say something reasonable that applies to KeyPackages, Proposals, and Commits without tripping over myself was a challenge.